scheduler/replica_scheduler.go

```diff
@@ -569,3 +570,55 @@ func (rcs *ReplicaScheduler) GetDiskSchedulingInfo(disk types.DiskSpec, diskStat
 	}
 	return info, nil
 }
+
+func (rcs *ReplicaScheduler) CheckReplicasSizeExpansion(v *longhorn.Volume, newSize int64) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "error while CheckReplicasSizeExpansion for volume %v", v.Name)
+	}()
+
+	replicas, err := rcs.ds.ListVolumeReplicas(v.Name)
+	if err != nil {
+		return err
+	}
+	diskIDToReplicaCount := map[string]int64{}
+	diskIDToDiskInfo := map[string]*DiskSchedulingInfo{}
+	for _, r := range replicas {
+		if r.Spec.NodeID == "" {
+			continue
+		}
+		node, err := rcs.ds.GetNode(r.Spec.NodeID)
+		if err != nil {
+			return err
+		}
+		diskSpec, diskStatus, ok := findDiskSpecAndDiskStatusInNode(r.Spec.DiskID, node)
+		if !ok {
+			return fmt.Errorf("cannot find the disk %v in node %v", r.Spec.DiskID, node.Name)
+		}
+		diskInfo, err := rcs.GetDiskSchedulingInfo(diskSpec, &diskStatus)
+		if err != nil {
+			return fmt.Errorf("failed to GetDiskSchedulingInfo %v", err)
+		}
+		diskIDToDiskInfo[r.Spec.DiskID] = diskInfo
+		diskIDToReplicaCount[r.Spec.DiskID] = diskIDToReplicaCount[r.Spec.DiskID] + 1
+	}
+
+	delSize := newSize - v.Spec.Size
+	for diskID, diskInfo := range diskIDToDiskInfo {
+		requestingSizeExpansion := delSize * diskIDToReplicaCount[diskID]
+		if !rcs.IsSchedulableToDisk(requestingSizeExpansion, 0, diskInfo) {
+			logrus.Errorf("Cannot expand replica size from %v to %v on disk %v with %+v", v.Spec.Size, newSize, diskID, diskInfo)
+			return fmt.Errorf("cannot expand replica size from %v to %v on disk %v with %+v ", v.Spec.Size, newSize, diskID, diskInfo)
+		}
+	}
+	return nil
+}
+
+func findDiskSpecAndDiskStatusInNode(diskID string, node *longhorn.Node) (types.DiskSpec, types.DiskStatus, bool) {
```

Comment:
NIT: To avoid confusion, please use `diskUUID` here and name the disk map key as `diskName`.

Reply:
Thank you! I fixed the names

Like @innobead mentioned, I think we assume that users don't mess up with the CRs directly otherwise we would have a lot of problems not just with volume expansion.